tokio/src/runtime/task/harness.rs

```diff
@@ -343,6 +355,15 @@ where
             }));
         }
 
+        if snapshot.is_join_interested() && snapshot.is_join_waker_set() {
+            // If JOIN_INTEREST and JOIN_WAKER are still set at this point, the runtime should
+            // drop the join waker as the join handle is not allowed to modify the waker
+            // following rule 6 in task/mod.rs
+            unsafe {
+                self.trailer().set_waker(None);
+            }
```

Comment:
I don't think this works. The join handle may not be allow to modify the waker, but it *is* allowed to read from it — which it does to call `will_wake`. This can data race with such calls.

Reply:
Oh yeah I see the problem. I attempted to fix it.